### ISSUE TITLE
Deadlock when MockScheduler.scheduleOnce and VirtualTime.advance are called at the same time

### ISSUE THREAD

**Issue by emrecelikten** [miguno/akka-mock-scheduler#4]
Hello,

If MockScheduler.scheduleOnce and VirtualTime.advance are called at the same time, I think the following deadlock occurs:

MockScheduler locks its own lock in MockScheduler.addToTasks, attempts to access time.elapsed which requires a lock in VirtualTime; but at the same, VirtualTime.advance locks its own lock, then attempts MockScheduler.tick, which requires a lock in MockScheduler.

This causes tests to freeze sporadically.

See the following stack trace for an example:

```
"ScalaTest-run-running-FooSpec" #1 prio=5 os_prio=31 tid=0x00007f947a002800 nid=0x1903 waiting for monitor entry [0x000000010c558000]
   java.lang.Thread.State: BLOCKED (on object monitor)
    at com.miguno.akka.testing.MockScheduler.tick(MockScheduler.scala:38)
    - waiting to lock <0x000000079763d0a8> (a java.lang.Object)
    at com.miguno.akka.testing.VirtualTime.advance(VirtualTime.scala:34)
    - locked <0x000000079763d158> (a java.lang.Object)

"TestSystem-akka.actor.default-dispatcher-3" #14 prio=5 os_prio=31 tid=0x00007f9479a02800 nid=0x6303 waiting for monitor entry [0x000000012705a000]
   java.lang.Thread.State: BLOCKED (on object monitor)
    at com.miguno.akka.testing.VirtualTime.elapsed(VirtualTime.scala:20)
    - waiting to lock <0x000000079763d158> (a java.lang.Object)
    at com.miguno.akka.testing.MockScheduler.addToTasks(MockScheduler.scala:60)
    - locked <0x000000079763d0a8> (a java.lang.Object)
    at com.miguno.akka.testing.MockScheduler.scheduleOnce(MockScheduler.scala:51)
```

**Comment by miguno** [comment]
@emrecelikten Many thanks for reporting this issue.  Could you test-drive the fix in https://github.com/miguno/akka-mock-scheduler/commit/546891862fcae177d19dd0837b088482d8e6faa7?

**Comment by emrecelikten** [comment]
Thanks for the quick fix! I have run my test suite more than 50 times, no freezes but sporadic failures due to race conditions at times where the deadlocks would occur. So I think it's fine now.

**Comment by miguno** [comment]
Could you elaborate on the sporadic failures?

**Comment by emrecelikten** [comment]
Sorry, I meant to say that my tests were not passing due to race conditions in my test code. I am testing an Akka actor (which runs in another thread) - after receiving a message, the actor immediately schedules a message to send after some time using scheduleOnce. In the test, I am checking if the message is actually sent in the expected time. The race condition happens when I do something like

``` scala
actor ! SomeMessage
time.advance(5.seconds)
expectMsg(SomeResponse)
```

So, actor schedules the message either before or after time.advance is called - there is a race since the test code and the actor code is being run on different threads. If former, the test works as expected. If latter, then the actor schedules the message after 5 seconds were already elapsed, so expectMsg gets nothing since the scheduler should still wait for another 5 seconds to fire the message.

Before your fix, the code above caused deadlocks as both threads attempted to lock the objects. Now it is just a race condition which sometimes causes the test to fail.

I doubt there is anything this (amazing) library can do for this type of problem. I implemented a temporary fix (i.e. hack) by wrapping the advance function in a sleepAdvance function which does some very small amount of Thread.sleep() before trying to advance the time, so the actor manages to schedule first. (or putting expectNoMsg() after sending, which would also be the correct approach for the test above)

**Comment by miguno** [comment]
Ah, thanks for the clarification.

FYI: I released the fixed version as `0.3.1`.  It may take a few hours though until it's synced to / available in Maven Central.

**Comment by emrecelikten** [comment]
Great, thanks again!
